### PR TITLE
[JENKINS-45522] Introduced ClassFilter.appendDefaultFilter

### PR DESCRIPTION
**Pull request by jglick**
[JENKINS-45522](https://issues.jenkins-ci.org/browse/JENKINS-45522)

Allows us to get rid of some awful reflection in core.

@reviewbybees

### CODE REVIEW

File: src/main/java/hudson/remoting/ClassFilter.java [hunk 1]
```diff
@@ -89,6 +89,17 @@ public final Class check(Class c) {
     public static final ClassFilter DEFAULT = createDefaultInstance();
 
     /**
+     * Adds an additional exclusion to {@link #DEFAULT}.
+     * Does nothing if the default list has already been customized via {@link #FILE_OVERRIDE_LOCATION_PROPERTY}.
+     * @param filter a regular expression for {@link Class#getName} which, if matched according to {@link Matcher#matches}, will blacklist the class
+     */
+    public static void appendDefaultFilter(Pattern filter) {
+        if (System.getProperty(FILE_OVERRIDE_LOCATION_PROPERTY) == null) {
+            ((RegExpClassFilter) DEFAULT).add(filter.toString());
```
Comment (oleg-nenashev) [inline src/main/java/hudson/remoting/ClassFilter.java]: seems to be a leaky API presumption. FindBugs will grumble once enabled. Add class check and checked exception at least 🐜

File: src/main/java/hudson/remoting/ClassFilter.java [hunk 2]
```diff
@@ -89,6 +89,17 @@ public final Class check(Class c) {
     public static final ClassFilter DEFAULT = createDefaultInstance();
 
     /**
+     * Adds an additional exclusion to {@link #DEFAULT}.
+     * Does nothing if the default list has already been customized via {@link #FILE_OVERRIDE_LOCATION_PROPERTY}.
+     * @param filter a regular expression for {@link Class#getName} which, if matched according to {@link Matcher#matches}, will blacklist the class
+     */
+    public static void appendDefaultFilter(Pattern filter) {
```
Comment (oleg-nenashev) [inline src/main/java/hudson/remoting/ClassFilter.java]: Throw checked exceptions to be able to propagate errors if we need to add some (like the class assumption failure below)

File: src/main/java/hudson/remoting/ClassFilter.java [hunk 3]
```diff
@@ -89,6 +89,17 @@ public final Class check(Class c) {
     public static final ClassFilter DEFAULT = createDefaultInstance();
 
     /**
+     * Adds an additional exclusion to {@link #DEFAULT}.
+     * Does nothing if the default list has already been customized via {@link #FILE_OVERRIDE_LOCATION_PROPERTY}.
+     * @param filter a regular expression for {@link Class#getName} which, if matched according to {@link Matcher#matches}, will blacklist the class
+     */
```
Comment (oleg-nenashev) [inline src/main/java/hudson/remoting/ClassFilter.java]: since TODO

File: src/main/java/hudson/remoting/ClassFilter.java (hunk 1 above)
Comment (jglick) [inline src/main/java/hudson/remoting/ClassFilter.java]: I do not believe there is any need to check the cast. `DEFAULT` cannot be anything other than a `RegExpClassFilter`. Could adjust return type of `createDefaultInstance` to make this more apparent.

File: src/main/java/hudson/remoting/ClassFilter.java (hunk 2 above)
Comment (jglick) [inline src/main/java/hudson/remoting/ClassFilter.java]: What checked exceptions? I see no need for any.

**Comment by jglick** [comment]
I have responded to all comments and do not believe they are real issues.

**Comment by oleg-nenashev** [comment]
@jglick IIRC we have discussed it during one of the dailys ~1 month ago. I believe we need a way to propagate errors from `appendDefaultFilter()` if we need to harden the implementation (e.g. to add blacklists). If you do not want to change the implementation, let me know. But I keep 🐛 for the current state

**Comment by jglick** [comment]
> if we need to harden the implementation (e.g. to add blacklists)

What does this mean? The API’s sole purpose is to blacklist entries.

> If you do not want to change the implementation

I do not follow what you are asking for or why.

**Comment by oleg-nenashev** [comment]
Created #185 with my vision of this change

**Pull request merged**